trieste/utils/pareto.py

```diff
@@ -101,10 +108,155 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+            lower_result, upper_result = tf.cond(
+                test_accepted,
+                lambda: self._accepted_test_body(lower_result, upper_result, lower_idx, upper_idx),
+                lambda: (lower_result, upper_result),
+            )
+
+            test_rejected = self._is_test_required((lower + jitter) < front)
+            divide_conquer_cells = tf.cond(
+                tf.logical_and(test_rejected, tf.logical_not(test_accepted)),
+                lambda: self._rejected_test_body(
+                    cell, lower, upper, divide_conquer_cells, total_size, threshold
+                ),
+                lambda: divide_conquer_cells,
+            )
+
+            return divide_conquer_cells, lower_result, upper_result
+
+        _, lower_result, upper_result = tf.while_loop(
+            lambda divide_conquer_cells, lower_result, upper_result: divide_conquer_cells.shape[0]
+            > 0,
+            while_body,
+            loop_vars=[divide_conquer_cells, lower_result, upper_result],
+            shape_invariants=[
+                tf.TensorShape([None, 2, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+            ],
+        )
+        return BoundedVolumes(lower_result, upper_result)
+
+    def _accepted_test_body(self, lower_result, upper_result, lower_idx, upper_idx):
+        lower_result = tf.concat([lower_result, lower_idx[None]], axis=0)
+        upper_result = tf.concat([upper_result, upper_idx[None]], axis=0)
+        return lower_result, upper_result
+
+    def _rejected_test_body(
+        self,
+        cell: TensorType,
+        lower: TensorType,
+        upper: TensorType,
+        divide_conquer_cells: TensorType,
+        total_size: TensorType,
+        threshold: TensorType,
+    ):
+
+        divide_conquer_cells_dist = cell[1] - cell[0]
+        hc_size = tf.math.reduce_prod(upper - lower, axis=0, keepdims=True)
+
+        not_unit_cell = tf.reduce_any(divide_conquer_cells_dist > 1)
+        vol_above_thresh = tf.reduce_all((hc_size[0] / total_size) > threshold)
+        divide_conquer_cells = tf.cond(
+            tf.logical_and(not_unit_cell, vol_above_thresh),
+            lambda: self._divide_body(divide_conquer_cells, divide_conquer_cells_dist, cell),
+            lambda: tf.identity(divide_conquer_cells),
+        )
+        return divide_conquer_cells
+
+    def _divide_body(
+        self,
+        divide_conquer_cells: TensorType,
+        divide_conquer_cells_dist: TensorType,
+        cell: TensorType,
+    ):
+        edge_size, idx = tf.reduce_max(divide_conquer_cells_dist), tf.argmax(
+            divide_conquer_cells_dist
+        )
+        edge_size1 = int(tf.round(tf.cast(edge_size, dtype=tf.float32) / 2.0))
+        edge_size2 = edge_size - edge_size1
+
+        upper = tf.unstack(tf.identity(cell[1]))
+        upper[idx] = upper[idx] - edge_size1
+        upper = tf.stack(upper)
```

Comment:
this mutation is probably not going to play nice with `tf.function`. Is it possible to make `edge_size1` into an appropriately shaped tensor and subtract it straight from `tf.identity(cell[1])`?

Reply:
similar for `lower` later

Reply:
is there rank1 version of `tf.sparse.SparseTensor` ?

Reply:
I don't know, but this might work (with some tweaking)
```
tf.concat(tf.zeros([idx, ...]), edge_size1 * tf.ones([...]), tf.zeros([len(cell[1]) - idx, ...]), axis=...)
```

Reply:
I'll try it

Reply:
tried it [commit](https://github.com/secondmind-labs/trieste/pull/207/commits/b1c1b64c6136f26929fcd7187a63c40c0bc2d18b) what do you think?